glean-core/rlb/examples/delayed-ping-data.rs

```diff
@@ -0,0 +1,126 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::env;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::PathBuf;
+
+use once_cell::sync::Lazy;
+
+use flate2::read::GzDecoder;
+use glean::net;
+use glean::{private::PingType, ClientInfoMetrics, ConfigurationBuilder};
+
+pub mod glean_metrics {
+    use glean::{private::CounterMetric, CommonMetricData, Lifetime};
+
+    #[allow(non_upper_case_globals)]
+    pub static sample_counter: once_cell::sync::Lazy<CounterMetric> =
+        once_cell::sync::Lazy::new(|| {
+            CounterMetric::new(CommonMetricData {
+                name: "sample_counter".into(),
+                category: "test.metrics".into(),
+                send_in_pings: vec!["prototype".into()],
+                disabled: false,
+                lifetime: Lifetime::Ping,
+                ..Default::default()
+            })
+        });
+}
+
+#[derive(Debug)]
+struct MovingUploader(PathBuf);
+
+impl MovingUploader {
+    fn new(mut path: PathBuf) -> Self {
+        path.push("sent_pings");
+        std::fs::create_dir_all(&path).unwrap();
+        Self(path)
+    }
+}
+
+impl net::PingUploader for MovingUploader {
+    fn upload(&self, upload_request: net::PingUploadRequest) -> net::UploadResult {
+        let net::PingUploadRequest {
+            body, url, headers, ..
+        } = upload_request;
+        let mut gzip_decoder = GzDecoder::new(&body[..]);
+        let mut s = String::with_capacity(body.len());
+
+        let data = gzip_decoder
+            .read_to_string(&mut s)
+            .ok()
+            .map(|_| &s[..])
+            .or_else(|| std::str::from_utf8(&body).ok())
+            .unwrap();
+
+        let docid = url.rsplit('/').next().unwrap();
+        let out_path = self.0.join(format!("{docid}.json"));
+        let mut fp = File::create(out_path).unwrap();
+
+        // pseudo-JSON, let's hope this works.
+        writeln!(fp, "{{").unwrap();
+        writeln!(fp, "  \"url\": {url},").unwrap();
+        for (key, val) in headers {
+            writeln!(fp, "  \"{key}\": \"{val}\",").unwrap();
+        }
+        writeln!(fp, "}}").unwrap();
+        writeln!(fp, "{data}").unwrap();
+
+        net::UploadResult::http_status(200)
+    }
+}
+
+#[allow(non_upper_case_globals)]
+pub static PrototypePing: Lazy<PingType> =
+    Lazy::new(|| PingType::new("prototype", true, true, false, true, true, vec![], vec![]));
+
+fn main() {
+    env_logger::init();
+
+    let mut args = env::args().skip(1);
+
+    let data_path = PathBuf::from(args.next().expect("need data path"));
+    let state: u32 = args.next().as_deref().unwrap_or("0").parse().unwrap();
```

Comment:
I think I'd prefer it if `state` were a string: "accumulate_one_and_pretend_crash", "accumulate_ten_and_orderly_shutdown", "submit_ping"